src/EditorFeatures/Core/Implementation/CommentSelection/ToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,456 @@
+            if (!experimentationService.IsExperimentEnabled(WellKnownExperimentNames.RoslynToggleBlockComment))
+            {
+                return emptyResult;
+            }
+
+            var commentInfo = await service.GetInfoAsync(document, selectedSpans.First().Span.ToTextSpan(), cancellationToken).ConfigureAwait(false);
+            if (commentInfo.SupportsBlockComment)
+            {
+                return await ToggleBlockComments(document, commentInfo, selectedSpans, cancellationToken).ConfigureAwait(false);
+            }
+
+            return emptyResult;
+        }
+
+        /// <summary>
+        /// Retrieves data about the commented selection that can be optionally overriden by subclasses.
+        /// </summary>
+        protected virtual async Task<IToggleBlockCommentDocumentDataProvider> GetBlockCommentDocumentData(Document document, ITextSnapshot snapshot,
+            CommentSelectionInfo commentInfo, CancellationToken cancellationToken)
+        {
+            return new ToggleBlockCommentDocumentDataProvider(snapshot, commentInfo);
+        }
+
+        private async Task<CommentSelectionResult> ToggleBlockComments(Document document, CommentSelectionInfo commentInfo,
+            NormalizedSnapshotSpanCollection selectedSpans, CancellationToken cancellationToken)
+        {
+            var blockCommentDataProvider = await GetBlockCommentDocumentData(document, selectedSpans.First().Snapshot, commentInfo, cancellationToken).ConfigureAwait(false);
+
+            var blockCommentedSpans = blockCommentDataProvider.GetBlockCommentsInDocument();
+            var blockCommentSelections = selectedSpans.Select(span => new BlockCommentSelectionHelper(blockCommentedSpans, span)).ToList();
+
+            var returnOperation = Operation.Uncomment;
+
+            var uncommentChanges = new List<TextChange>();
+            var uncommentTrackingSpans = new List<CommentTrackingSpan>();
+            // Try to uncomment until an uncommented span is found.
+            foreach (var blockCommentSelection in blockCommentSelections)
+            {
+                // If any selection does not have comments to remove, then the operation should be comment.
+                if (!TryUncommentBlockComment(blockCommentedSpans, blockCommentSelection, uncommentChanges, uncommentTrackingSpans, commentInfo))
+                {
+                    returnOperation = Operation.Comment;
+                    break;
+                }
+            }
+
+            if (returnOperation == Operation.Comment)
+            {
+                var commentChanges = new List<TextChange>();
+                var commentTrackingSpans = new List<CommentTrackingSpan>();
+                blockCommentSelections.ForEach(
+                    blockCommentSelection => BlockCommentSpan(blockCommentSelection, blockCommentDataProvider, commentChanges, commentTrackingSpans, commentInfo));
```

Comment:
this is pretty unclear that ths is mutating.  ideally just use a real foreach here.